src/Raven.Server/Documents/Handlers/BatchHandler.cs

```diff
@@ -411,6 +411,16 @@ private static List<Index> GetImpactedIndexesToWaitForToBecomeNonStale(DocumentD
                 }
             }
             return indexesToCheck;
+
+            bool IndexContainsModifiedCollections(Index index)
+            {
+                return index.GetReferencedCollections().Any(kvp =>
+                {
+                    var collections = kvp.Value.Select(x => x.Name).ToHashSet();
```

Comment:
This will allocate (per each SaveChanges). Better to introduce virtual method in Index e.g. Index.WorksOnAnyCollection(collections). Implement it in base and override in for indexes that have references collections (the ones that override the GetReferencedCollections).